### ISSUE TITLE
Refactor internals based on ExtensionArray interface

### ISSUE THREAD

**Issue by jorisvandenbossche** [geopandas/geopandas#1000]
For the next release (0.6.0), I want to focus on refactoring the internals using the pandas ExtensionArray interface (https://pandas-dev.github.io/pandas-blog/pandas-extension-arrays.html, https://pandas.pydata.org/pandas-docs/stable/development/extending.html#extension-types)

What does this entail:

* Converting the basic GeometryArray (https://github.com/geopandas/geopandas/pull/993) to a full ExtensionArray class (adding array operations like indexing, ...), and integrating this in the GeoSeries/GeoDataFrame classes -> this is being done in https://github.com/geopandas/geopandas/pull/835

This has some consequences:

* Drop support for pandas < 0.23
    * I first thought to have a compatibility layer to also support older pandas versions, but: this adds a lot of complexity, and will not be possible anyhow once we switch the GeometryArray to use cython / store pointers instead of shapely objects. Since the 0.6.0 release with GeometryArray in mainly a preparation for that, dropping old pandas versions already now seems like the better option.

It is also the perfect moment to discuss / re-evaluate some API aspects, and also enables some new functionality:

- 

We already have issue https://github.com/geopandas/geopandas/issues/680 that discusses some of this, but also explores the options to have *just* the extension array without the current subclasses we have in GeoPandas. 
In this issue I want to focus on the discussion items and consequences of starting to rely on the ExtensionArray interface, within the current GeoPandas structure.

**Comment by jorisvandenbossche** [comment]
cc @geopandas/collaborators

**Comment by jorisvandenbossche** [comment]
The initial refactor PR is merged: https://github.com/geopandas/geopandas/pull/835

Some follow-up issues from the initial PR:

- [ ] expand support of `fillna` (now only for scalar values)
- [ ] change the missing value handling? Let `align` result in missing values instead of empty geometries; change default `fillna` of filling with empty geometry; let `isna` ignore empty geometries
- [ ] remove some of the GeoSeries/DataFrame methods that override pandas ones that are no longer necessary

\+ the ideas from above.

**Comment by martinfleis** [comment]
> Strict GeoSeries / GeoDataFrame constructors?

I vote for this. For GeoSeries definitely, as we have a very clear definition "_A GeoSeries is essentially a vector where each entry in the vector is a set of shapes corresponding to one observation_". I would not return Series in this case, I would raise an error as you suggested. It feels much cleaner that expecting that user actually wanted Series.
Similarly for GeoDataFrame, definition is clear "_A GeoDataFrame is a tabular data structure that contains a GeoSeries_". If we drop geometry, it is not GeoDataFrame anymore (unless there is another column containing GeoSeries...). So, generally yes, return DataFrame after dropping geometry, but we should come up with a solution for situations where you still have GeoSeries in gdf even after drop. But there is another tricky issue, which is quite common and that is generating GeoDataFrame and then assigning geometry...

>  It will enable certain hashing-based functionalities, like the unique method, or grouping in groupby

Unless this will work as topological equality (which might be slow) I would not use it. Otherwise even LineStrings A-B and B-A would be seen as unequal, not mentioning multi-geometries. As you mentioned in #725, `GEOSNormalize` might solve this (so we can use WKB). Would it make sense to go through shapely? Adding PR allowing normalise there and then using it in geopandas? Or is it too convoluted?

> Sorting: do we want to enable sorting functionality?

How do you want to do it? Ignoring technical part, what would be sorting based on? Latitude? Longitude? Of which point? Area?

**Comment by jorisvandenbossche** [comment]
> But there is another tricky issue, which is quite common and that is generating GeoDataFrame and then assigning geometry...

Yes, I think that is indeed a problem if we would want to make GeoDataFrame strict. 
I think the pattern is something like `geopandas.GeoDataFrame(data_without_geometry).set_geometry(geometry_data)` (while people should rather do `geopandas.GeoDataFrame(data, geometry=geometry_data)` in this case). Is that correct?

But I agree that for GeoSeries it should be easier to be strict.

The main issue that should be checked is the `GeoSeries._constructor` which pandas uses to deal with subclasses (I need to check if that attribute is expected to be an actual class, or if a constructor that returns a class is sufficient).

> Unless this will work as topological equality (which might be slow) I would not use it. 

Yeah, so that is the problem with using WKB for this .. 
So basically it is like using `equals_exact` (although without a tolerance for errors) instead of `equals` (the topological equality method).

> Adding PR allowing normalise there and then using it in geopandas? Or is it too convoluted?

I think it is valuable for shapely anyway.

> > Sorting: do we want to enable sorting functionality?
>
> How do you want to do it? Ignoring technical part, what would be sorting based on? Latitude? Longitude? Of which point? Area?

Not really an idea :) Could be based on the text repr? (so for points it would be sorting on x, y) 
I am not really sure the exact way is important (as it will never be very sensible). The main reason I would consider adding that is for the ability to have a series sorted (eg to ensure a consistent order, to ensure duplicates are grouped together, ..) rather than for the actual order of the ordered series. 
But this is certainly not a priority ;)
